RFC: Reserved prefixes in the 2021 edition (#3101)
rust-lang/rfcs: text/0000-reserved_prefixes.md

1. It increases the amount of leeway for future language changes that might wish to use this space (e.g. a hypothetical mechanism for edition-specific keywords might be written as `edition2015#use`).
2. It has symmetry with the existing notion of [literal suffixes](https://doc.rust-lang.org/reference/tokens.html#suffixes).
3. It avoids complicating the grammar and parser with bespoke concepts.

Finally, this RFC also proposes that this same syntactic reservation be applied to string literals (including raw string literals), whose syntax was the original inspiration for this design space. Once again, this reservation would be mostly unobservable by end-users and would only manifest in code using macros like so:

```rust
macro_rules! demo {
    ( $a:tt ) => { println!("one token") };
    ( $a:tt $b:tt ) => { println!("two tokens") };
    ( $a:tt $b:tt $c:tt $d:tt ) => { println!("four tokens") };
}

demo!(br"foo");
demo!(bar"foo");
demo!(bar#"foo"#);
```

Prior to the 2021 edition, this produces:
```
one token
two tokens
four tokens
```

Following the 2021 edition, all would be lexed as one token. Once again, whitespace could be inserted to mitigate any breakage.

> The motivation in this case, aside from the symmetry with prefixed identifiers, would be to leave open the design space for additional string literal prefixes other than `b"` and `r"`, e.g. hypothetical format string literals `f"`, `String` literals `s"`, `CString` literals `c"`,  `OsString` literals `o"`, UTF-16 literals `w"`, user-overloadable string literals `x"`, etc., and any sensible combinations of these.

Reviewer: ```suggestion
The motivation in this case, aside from the symmetry with prefixed identifiers, would be to leave open the design space for additional string literal prefixes other than `b"` and `r"`. Some hypothetical examples (not necessarily planned features or planned syntax): format string literals `f"`, `String` literals `s"`, `CString` literals `c"`,  `OsString` literals `o"`, UTF-16 literals `w"`, user-overloadable string literals `x"`, etc., and any sensible combinations of these.
```

Let's carefully dodge any potential bikesheds that might happen over things that aren't actually proposals. I've seen RFCs derailed over casual mentions of future possibilities.

New tokenizer rules are introduced:

> RESERVED_IDENTIFIER : IDENTIFIER_OR_KEYWORD<sub>Except `r`</sub> `#` IDENTIFIER_OR_KEYWORD
>
> RESERVED_STRING_LITERAL : IDENTIFIER_OR_KEYWORD RAW_BYTE_STRING_LITERAL | IDENTIFIER_OR_KEYWORD BYTE_STRING_LITERAL | IDENTIFIER_OR_KEYWORD<sub>Except `b`</sub> RAW_STRING_LITERAL | IDENTIFIER_OR_KEYWORD<sub>Except `b`, `r`, `br`</sub> STRING_LITERAL

Any use of a reserved identifier or reserved string literal is a compilation error.

The use of "identifier" in this document should be taken to refer to the definition of "identifier" that is in use by Rust as of the 2021 edition. At the time of this writing, the `non_ascii_idents` feature is not yet stabilized, but is on track to be. If `non_ascii_idents` is stabilized before the 2021 edition, then the syntactic reservations that take place in the 2021 edition will include things like `über#foo`. However, if `non_ascii_idents` is *not* stabilized before the 2021 edition, then any subsequent stabilization of `non_ascii_idents` would need to take care to *not* expand the reservations in this RFC, and instead defer that task to the next edition.

> An edition migration will be implemented that looks for `ident#ident` or `ident"string"` within macro calls and inserts whitespace to force individual tokenization.

Reviewer: ```suggestion
An edition migration may be implemented that looks for `ident#ident` or `ident"string"` within macro calls and inserts whitespace to force individual tokenization.
```

We may or may not need such a migration, depending on what a crater run finds.

# Drawbacks
[drawbacks]: #drawbacks

* Complicates macro tokenizing rules.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

* Reserve only `ident#foo` and not `ident"foo"`. The former has a concrete RFC that would benefit from this, but the latter is currently just aspirational.

> * Instead of `ident`, reserve only `[a-z]+` or `[a-z]`. However, it would probably be even more surprising if `x#foo` were considered one token and `X#foo` or `xx#foo` were considered three.

Reviewer: ```suggestion
* Instead of `ident`, reserve only `[a-z]+` or `[a-z]`. However, it would probably be even more surprising if `x#foo` were considered one token and `X#foo` or `xx#foo` were considered three. In addition, reserving only `[a-z]` would force future language extensions to use exclusively pithy single-letter syntax, even for features that may not be common enough to warrant such abbreviated syntax. Reserving identifiers in this space provides more flexibility for future language design, without impacting Rust programs.
```

# Guide-level explanation
When writing macros that accept token trees as inputs, be aware that certain grammatical productions which have no meaning in Rust are nonetheless accepted by the grammar, as they represent "reserved space" for future Rust language development. This behavior can be observed by macros that consume token trees. In particular, anything of the form `<identifier>#<identifier>` or `<identifier>"<string contents>"` will be consumed by a macro as a single token tree. This is in contrast to, for example, `<identifier>@<identifier>`, which would be consumed as three token trees, as `@` does not indicate reserved syntax in the way that `#` does. These single tokens rely on the absence of whitespace, so a macro invocation can use `<identifier> # <identifer>` (note the spaces) as a way to consume individual tokens adjacent to a `#`.

> # Reference-level explanation

Reviewer 2: We need to specify what span of text is used to determine the edition. In this case, the change occurs in the lexer itself, so I believe the edition is simply the global edition for the current crate, and not one that is tied to any particular span. This is worth saying.

* Instead of `ident`, reserve only `[a-zA-Z_][a-zA-Z0-9_]*` (ASCII-only identifiers). This would cover the space future Rust language design extensions are likely to use. However, the explanation of the reserved space would require presenting a distinct concept separate from the definition of identifiers. In addition, reserving only ASCII identifiers seems unlikely to provide a benefit to future Rust programs.
* Instead of `ident`, reserve prefixes that permit any sequence of identifier continuation characters. This would allow things like preceding digits, e.g. `4#foo`.
* In addition to adding reserved prefixes to string literals, add them to numeric literals as well: `ident#1234`, `ident#56.78`.

#  Unresolved questions
[unresolved-questions]: #unresolved-questions

> * Is the automatic migration possible to implement? I'm unclear if rustfmt, and hence rustfix, is capable of peering inside of macro invocations.

Reviewer 2: This is definitely possible to implement. The relevant question is whether *rustc* is capable of peering inside macro invocation arguments, since it is the one responsible for issuing lints (it can).

# Drawbacks
[drawbacks]: #drawbacks

* Complicates macro tokenizing rules.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

* Reserve only prefixed identifiers, and not prefixed literals. The former has a concrete RFC that would benefit from this, but the latter is currently just aspirational.
* Instead of `ident`, reserve only `[a-z]+` or `[a-z]`. However, reserving only `[a-z]` would force future language extensions to use exclusively pithy single-letter syntax, even for features that may not be common enough to warrant such abbreviated syntax. Reserving identifiers in this space provides more flexibility for future language design, without impacting Rust programs.
* Instead of `ident`, reserve only `[a-zA-Z_][a-zA-Z0-9_]*`, the set of ASCII-only identifiers. This would cover the space future Rust language design extensions are likely to use. However, the explanation of the reserved space would require presenting a distinct concept separate from the definition of identifiers. In addition, reserving only ASCII identifiers seems unlikely to provide a benefit to future Rust programs.

> * Instead of `ident`, reserve prefixes that permit any sequence of identifier continuation characters. This would allow things like preceding digits, e.g. `4#foo`.

Reviewer 2: We should probably include a discussion of erroring in the lexer vs parser and why the RFC opted for the latter.
